client/auth/handle-emailed-link-form.jsx

```diff
@@ -0,0 +1,172 @@
+
+/**
+ * Provide the "Magic Login" token to get logged in (or get the credentials to do so ourself)
+ * @param  {object} data - object containing an email address & token
+ * @param  {Function} fn - Function to invoke when request is complete
+ * @returns {Promise} promise
+ */
+function postToMagicLoginTokenHandler( data ) {
+	data.client_id = config( 'wpcom_signup_id' );
+	data.client_secret = config( 'wpcom_signup_key' );
+
+	return request
+		.post( 'https://wordpress.com/wp-login.php?action=magic-login' )
+		.withCredentials()
+		.send( data )
+		.set( {
+			Accept: 'application/json',
+			'Content-Type': 'application/x-www-form-urlencoded',
+		} )
+		.end( responseHandler );
+}
+
+function responseHandler( error, response ) {
+	response = response || {};
+
+	const { status, body } = response;
+
+	if ( ! status || 400 <= status && status < 500 ) {
+		debug( 'unsuccessful status: ' + status );
+		window.location.replace( '/login/link-has-expired' );
+		return;
+	}
+
+	if ( error || ! ( body && body.data && body.success ) ) {
+		debug( 'error: ' + JSON.stringify( error ) );
+		debug( 'body: ' + JSON.stringify( body ) );
+
+		// @TODO This is most likely a server or network error, maybe show a notice instead?
+		window.location.replace( '/login/link-has-expired' );
+		return;
+	}
+
+	const { data } = body;
+	const { redirect_to } = data;
+
+	if ( redirect_to.match( /^(https:\/\/wordpress\.com|http:\/\/calypso\.localhost:3000)(\/|$)/ ) ) {
+		debug( 'redirecting: ' + redirect_to );
+		window.location.replace( redirect_to );
+	} else {
+		window.location.replace( '/' );
+	}
+}
+
+class HandleEmailedLinkForm extends React.Component {
+	state = {
+		hasSubmitted: false,
+	};
+
+	handleSubmit = event => {
+		event.preventDefault();
+
+		debug( 'form submitted!' );
+
+		this.setState( {
+			hasSubmitted: true,
+		} );
+
+		const postData = {
+			email: this.props.emailAddress,
+			token: this.props.token,
+			tt: this.props.tokenTime,
+		};
+
+		debug( '`POST`ing credentials to WPCOM', postData );
+
+		postToMagicLoginTokenHandler( postData );
+	};
+
+	componentWillMount() {
+		const { emailAddress, token, tokenTime } = this.props;
+
+		if ( emailAddress && emailValidator.validate( emailAddress ) && token && tokenTime ) {
+			return;
+		}
+
+		window.location.replace( '/login/link-has-expired' );
+	}
+
+	render() {
+		const { currentUser, emailAddress, translate } = this.props;
+		const action = (
+			<Button primary disabled={ !! this.state.hasSubmitted } onClick={ this.handleSubmit }>
+				{ translate( 'Finish Login' ) }
+			</Button>
+		);
+		const title =
+			this.props.clientId === config( 'wpcom_signup_id' )
+				? translate( 'Continue to WordPress.com' )
+				: translate( 'Continue to WordPress.com on your WordPress app' );
+		const line = [
+			translate(
+				'Logging in as %(emailAddress)s', {
+					args: {
+						emailAddress,
+					}
+				}
+			)
+		];
+
+		if ( currentUser && currentUser.username ) {
```

Comment:
If `emailAddress` belongs to `currentUser`, should we just redirect? This would happen if the user clicked the link in the email again after they were already logged in.